src/gui/torrentcontenttreeview.cpp

```diff
@@ -107,6 +108,78 @@ void TorrentContentTreeView::keyPressEvent(QKeyEvent *event)
     }
 }
 
+void TorrentContentTreeView::setupDownloadPriorityMenu(QMenu *menu, const bool createSubMenu) const
+{
+    const auto model = qobject_cast<TorrentContentFilterModel *>(this->model());
+    Q_ASSERT(model);
+
+    const auto getSelectedRows = [this]() { return selectionModel()->selectedRows(TorrentContentModelItem::COL_PRIO); };
+
+    const auto applyPriorities = [&](const BitTorrent::DownloadPriority priority)
+    {
+        return [model, getSelectedRows, priority]()
+        {
+            for (const QModelIndex &index : asConst(getSelectedRows()))
+            {
+                model->setData(index, static_cast<int>(priority));
+            }
+            emit model->filteredFilesChanged();
+        };
+    };
+    const auto applyPrioritiesByOrder = [model, getSelectedRows]()
+    {
+        // Equally distribute the selected items into groups and for each group assign
+        // a download priority that will apply to each item. The number of groups depends on how
+        // many "download priority" are available to be assigned
+
+        const QModelIndexList selectedRows = getSelectedRows();
+        const qsizetype priorityGroups = 3;
+        const auto priorityGroupSize = std::max<qsizetype>(selectedRows.length() / priorityGroups, 1);
+
+        for (qsizetype i = 0; i < selectedRows.length(); ++i)
+        {
+            auto priority = BitTorrent::DownloadPriority::Ignored;
+            switch (i / priorityGroupSize)
+            {
+            case 0:
+                priority = BitTorrent::DownloadPriority::Maximum;
+                break;
+            case 1:
+                priority = BitTorrent::DownloadPriority::High;
+                break;
+            default:
+            case 2:
+                priority = BitTorrent::DownloadPriority::Normal;
+                break;
+            }
+
+            model->setData(selectedRows[i], static_cast<int>(priority));
+        }
+
+        emit model->filteredFilesChanged();
+    };
+
+    if (createSubMenu)
+    {
+        menu = menu->addMenu(tr("Priority"));
+        menu->addAction(tr("Do not download"), menu, applyPriorities(BitTorrent::DownloadPriority::Ignored));
+        menu->addAction(tr("Normal"),          menu, applyPriorities(BitTorrent::DownloadPriority::Normal));
+        menu->addAction(tr("High"),            menu, applyPriorities(BitTorrent::DownloadPriority::High));
+        menu->addAction(tr("Maximum"),         menu, applyPriorities(BitTorrent::DownloadPriority::Maximum));
+        menu->addSeparator();
+        menu->addAction(tr("By shown file order"), menu, applyPrioritiesByOrder);
+    }
+    else
+    {
+        menu->addAction(tr("Do not download"),  menu, applyPriorities(BitTorrent::DownloadPriority::Ignored));
+        menu->addAction(tr("Normal priority"),  menu, applyPriorities(BitTorrent::DownloadPriority::Normal));
```

Comment:
I don't really like it, the additional "... priority" part. It creates unnecessary more code, more work for translation and doesn't look good on screen for me (subjectively).
IMO for this kind of change it would be best to split into another PR for discussion.

Reply:
>I don't really like it, the additional "... priority" part. 

But will it look clear if "priority" is not mentioned at all?

P.S. Personally, I would not have any problems if it remained as it was previously, i.e. it was always displayed as a "Priority" submenu.

Reply:
> I don't really like it, the additional "... priority" part. It creates unnecessary more code, more work for translation and doesn't look good on screen for me (subjectively). IMO for this kind of change it would be best to split into another PR for discussion.

Seems I mistaken this as a new change... disregard it.

>I would not have any problems if it remained as it was previously, i.e. it was always displayed as a "Priority" submenu.

I would prefer that too, but other users like it... how do we solve this kind of dilemma? more configurable options? that sounds a bit overkill for me...

Reply:
>more configurable options? that sounds a bit overkill for me...

For me too.
>how do we solve this kind of dilemma?

Let's leave it as it is for now.

Another option I can imagine is always provide complete context menu but disable inappropriate items (i.e. disable "rename" etc. when multiple files are selected). Then we'll always have the same menu layout so priority-related items are always under "Priority" submenu.

Reply:
>Let's leave it as it is for now.

OK.

>Another option I can imagine is always provide complete context menu but disable inappropriate items (i.e. disable "rename" etc. when multiple files are selected). Then we'll always have the same menu layout so priority-related items are always under "Priority" submenu.

I think we should start enforcing this idea (at least to the content tree view) after this PR.